### PR TITLE
 RLP target gateways

### PR DESCRIPTION
**Pull request by eguzki**
### What

Fixes https://github.com/Kuadrant/kuadrant-controller/issues/160
Depends on https://github.com/Kuadrant/wasm-shim/pull/18

- [x] Allow RLP target reference to gateways
- [x]  New `RateLimitPolicy` struct according to the [proposal doc](https://github.com/Kuadrant/kuadrant-controller/blob/main/doc/proposals/rlp-target-gateway-resource.md#schema-crd-of-the-ratelimitpolicy)
```yaml
---
apiVersion: apim.kuadrant.io/v1alpha1
kind: RateLimitPolicy
metadata:
  name: my-rate-limit-policy
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute / Gateway
    name: myroute / mygateway
  rateLimits:
    - operations:
        - paths: ["/admin/*"]
          methods: ["GET"]
          hosts: ["example.com"]
      actions:
        - generic_key:
            descriptor_key: admin
            descriptor_value: "yes"
      limits:
        - conditions: ["admin == yes"]
          maxValue: 500
          seconds: 30
          variables: []
```
- [x] Limits reconciler
- [x] Gateway RLP reference list annotation reconciler
- [x] WASM Plugins reconciler
- [x] Limitador cluster Envoy Filter reconciler
- [x] Doc, user guides
- [ ] Tests (unit and e2e)

Robust reconciliation features. The kuadrant controller supports the following use cases:
- RLP object changes target to another  HTTPRoute
- RLP object changes target from HTTPRoute to Gateway
- HTTPRoute object changes parentRefs (references to parent Gateways)
- RLP object targeting HTTPRoute is deleted
- RLP object targeting Gateway is deleted
- HTTPRoute object changes `hostnames` list values
- RLP object changes spec: 
	- add new limit(s) / rule(s) / configuration(s)
	- remove limit(s) / rule(s) / configuration(s)
	- update limit(s)  / rule(s) / configuration(s) values
- Supports multiple RLP's with exactly the same `hostnames`

**NOTE**: this PR adds many changes into the rate limit policy controller. The reconciliation features required a heavy code refactor. It is highly recommended to review commit-by-commit for a better understanding of the changes being applied.  

### Verification steps
* deploy kuadrant locally
```
make local-setup
```
* deploy toystore service
```
kubectl apply -f examples/toystore/toystore.yaml
```
Create `toystore` HTTPRoute to configure routing to the toystore service

```yaml
kubectl apply -f - <<EOF
---
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: toystore
  labels:
    app: toystore
spec:
  parentRefs:
    - name: kuadrant-gwapi-gateway
      namespace: kuadrant-system
  hostnames: ["*.toystore.com"]
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: "/toy"
          method: GET
        - path:
            type: PathPrefix
            value: "/free"
          method: GET
        - path:
            type: Exact
            value: "/admin/toy"
          method: POST
      backendRefs:
        - name: toystore
          port: 80
EOF
```

Check `toystore` HTTPRoute works and it is not rate limited.

`GET /toy`: no rate limiting
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

`POST /admin/toy`: no rate limiting
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" -X POST http://localhost:9080/admin/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

### Rate limiting `toystore` HTTPRoute traffic

RateLimitPolicy applied for the `toystore` HTTPRoute.

| Endpoints | Rate Limits |
| ------------- | -----: |
| `POST /admin/toy` | **5** reqs / **10** secs (0.5 rps) |
| `GET /toy` | **8** reqs / **10** secs (0.8 rps) |
| `*` | **30** reqs / **10** secs (3.0 rps) |

```yaml
kubectl apply -f - <<EOF
---
apiVersion: apim.kuadrant.io/v1alpha1
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  rateLimits:
    - rules:
        - paths: ["/admin/toy"]
          methods: ["POST"]
      configurations:
        - actions:
            - generic_key:
                descriptor_key: admin-operation
                descriptor_value: "1"
      limits:
        - conditions:
            - "admin-operation == 1"
          maxValue: 5
          seconds: 10
          variables: []
    - rules:
        - paths: ["/toy"]
          methods: ["GET"]
      configurations:
        - actions:
            - generic_key:
                descriptor_key: get-operation
                descriptor_value: "1"
      limits:
        - conditions:
            - "get-operation == 1"
          maxValue: 8
          seconds: 10
          variables: []
    - configurations:
        - actions:
            - generic_key:
                descriptor_key: toystore
                descriptor_value: "1"
      limits:
        - conditions: ["toystore == 1"]
          maxValue: 30
          seconds: 10
          variables: []
EOF
```

Validating the rate limit policy.

`GET /toy` @ **1** rps (expected to be rate limited @ **8** reqs / **10** secs (0.8 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

`POST /admin/toy` @ **1** rps (expected to be rate limited @ **5** reqs / **10** secs (0.5 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" -X POST http://localhost:9080/admin/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

### Rate limiting Gateway traffic

RateLimitPolicy applied for the Gateway.

| Policy | Rate Limits |
| ------------- | -----: |
| `POST /*` | **2** reqs / **10** secs (0.2 rps) |
| Per remote IP | **25** reqs / **10** secs (2.5 rps) |

```yaml
kubectl apply -f - <<EOF
---
apiVersion: apim.kuadrant.io/v1alpha1
kind: RateLimitPolicy
metadata:
  name: kuadrant-gw
  namespace: kuadrant-system
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: kuadrant-gwapi-gateway
  rateLimits:
    - rules:
      - methods: ["POST"]
      configurations:
        - actions:
            - generic_key:
                descriptor_key: expensive-op
                descriptor_value: "1"
      limits:
        - conditions: ["expensive-op == 1"]
          maxValue: 2
          seconds: 10
          variables: []
    - configurations:
        - actions:
            - remote_address: {}
      limits:
        - conditions: []
          maxValue: 25
          seconds: 10
          variables: ["remote_address"]
EOF
```

Validating the rate limit policies (HTTPRoute and Gateway).

`GET /toy` @ **1** rps (expected to be rate limited @ **8** reqs / **10** secs (0.8 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

`POST /admin/toy` @ **1** rps (expected to be rate limited @ **2** reqs / **10** secs (0.2 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" -X POST http://localhost:9080/admin/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

Validating Gateway "Per Remote IP" policy

Stop all traffic.

`GET /free` @ **3** rps (expected to be rate limited @ **25** reqs / **10** secs (2.5 rps))

```
while :; do curl --write-out '%{http_code}\n' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/free -: --write-out '%{http_code}\n' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/free -: --write-out '%{http_code}\n' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/free | egrep --color "\b(429)\b|$"; sleep 1; done
```

### Remove Gateway's rate limit policy

```
kubectl delete ratelimitpolicy kuadrant-gw -n kuadrant-system
```

`GET /toy` @ **1** rps (expected to be rate limited @ **8** reqs / **10** secs (0.8 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

`POST /admin/toy` @ **1** rps (expected to be rate limited @ **5** reqs / **10** secs (0.5 rps))
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" -X POST http://localhost:9080/admin/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

### Remove HTTPRoute's rate limit policy

```
kubectl delete ratelimitpolicy toystore
```

`GET /toy`: no rate limiting
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" http://localhost:9080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

`POST /admin/toy`: no rate limiting
```
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: api.toystore.com" -X POST http://localhost:9080/admin/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

### CODE REVIEW

File: apis/apim/v1alpha1/ratelimitpolicy_types_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+func testBuildBasicGatewayRLP(name string) *RateLimitPolicy {
+	return testBuildBasicRLP(name, gatewayapiv1alpha2.Kind("Gateway"))
+}
+
+func testBuildBasicHTTPRouteRLP(name string) *RateLimitPolicy {
+	return testBuildBasicRLP(name, gatewayapiv1alpha2.Kind("HTTPRoute"))
+}
+
+// TestHelloName calls greetings.Hello with a name, checking
```
Comment (maleck13) [inline apis/apim/v1alpha1/ratelimitpolicy_types_test.go:41]: looks like copy paste left over

**Comment by eguzki** [comment]
The PR is ready for review. 

Currently, only tests are missing. I will implement them meanwhile the PR it is being reviewed

**Review by rahulanand16nov — APPROVED** [review]
Nice work! LGTM! Though I'd recommend a smaller PR :)

**Comment by eguzki** [comment]
> Though I'd recommend a smaller PR :)

I tried :'(

**Pull request merged**